python/ray/data/datasource/datasink.py

```diff
@@ -10,45 +10,25 @@
 logger = logging.getLogger(__name__)
 
 
-@dataclass
-@DeveloperAPI
-class WriteResult:
-    """Result of a write operation, containing stats/metrics
-    on the written data.
-
-    Attributes:
-        total_num_rows: The total number of rows written.
-        total_size_bytes: The total size of the written data in bytes.
-    """
-
-    num_rows: int = 0
-    size_bytes: int = 0
+# Generic type for the result of a write task.
+WriteResultType = TypeVar("WriteResultType")
 
-    @staticmethod
-    def aggregate_write_results(write_results: List["WriteResult"]) -> "WriteResult":
-        """Aggregate a list of write results.
 
-        Args:
-            write_results: A list of write results.
-
-        Returns:
-            A single write result that aggregates the input results.
-        """
-        total_num_rows = 0
-        total_size_bytes = 0
-
-        for write_result in write_results:
-            total_num_rows += write_result.num_rows
-            total_size_bytes += write_result.size_bytes
+@dataclass
+@DeveloperAPI
+class WriteResult(Generic[WriteResultType]):
+    """Aggregated result of the Datasink write operations."""
 
-        return WriteResult(
-            num_rows=total_num_rows,
-            size_bytes=total_size_bytes,
-        )
+    # Total number of written rows.
+    num_rows: int
+    # Total size in bytes of written data.
+    size_bytes: int
+    # Results of all `Datasink.write`.
+    write_task_results: List[WriteResultType]
```

Comment:
Nit: To avoid confusion between the `WriteResult` dataclass and the object returned from write tasks, it might clearer if we rename `write_task_results` to `write_return_types` (and `WriteResultType` to `WriteReturnType`)
```suggestion
    write_task_returns: List[WriteReturnType]
```